### ISSUE TITLE
Curiously recursive Protocol

### ISSUE THREAD

**Issue by jgarvin** [microsoft/pyright#5419]
**Describe the bug**
I don't know if this is intended. Pyright rejects this code, which Mypy accepts even with `--strict`:

```
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from typing import Protocol, TypeVar

T = TypeVar("T", covariant=True)

class Foo(Protocol[T]): ...

class Bar(Foo[Bar]): ...
```

Output:

```
$ pyright /tmp/type_test19.py 
/tmp/type_test19.py
  /tmp/type_test19.py:10:15 - error: "Bar" is not defined (reportUndefinedVariable)
1 error, 0 warnings, 0 informations 
```

The problem is `Bar` passing itself to `Foo`. I don't see why this should be disallowed, could be useful for expressing things like guaranteeing types have a factory method like:

```
class Foo(Protocol[T]):
    @staticmethod
    def new() -> T: ...
```

**Comment by debonte** [comment]
When specifying the base class you need to put `Bar` in quotes since it isn't defined yet. See https://peps.python.org/pep-0484/#forward-references.

```python
class Bar(Foo["Bar"]): ...
```
